File: UIKit and Animations/WWDC 2010/Designing Apps with Scroll Views/SampleCode/LargeImageDownsizing/LargeImageDownsizing/LargeImageDownsizing/Tiling.m
////
////  Tiling.m
////  LargeImage
////
////  Created by Seyed Samad Gholamzadeh on 1/12/18.
////
//
//#import <Foundation/Foundation.h>
//
//- (void)saveTilesOfSize:(CGSize)size
//forImage:(UIImage*)image
//toDirectory:(NSString*)directoryPath
//usingPrefix:(NSString*)prefix
//{
//	CGFloat cols = [image size].width / size.width;
//	CGFloat rows = [image size].height / size.height;
//	
//	int fullColumns = floorf(cols);
//	int fullRows = floorf(rows);
//	
//	CGFloat remainderWidth = [image size].width -
//	(fullColumns * size.width);
//	CGFloat remainderHeight = [image size].height -
//	(fullRows * size.height);
//	
//	
//	if (cols > fullColumns) fullColumns++;
//	if (rows > fullRows) fullRows++;
//	
//	CGImageRef fullImage = [image CGImage];
//	
//	for (int y = 0; y < fullRows; ++y) {
//		for (int x = 0; x < fullColumns; ++x) {
//			CGSize tileSize = size;
//			if (x + 1 == fullColumns && remainderWidth > 0) {
//					// Last column
//				tileSize.width = remainderWidth;
//			}
//			if (y + 1 == fullRows && remainderHeight > 0) {
//					// Last row
//				tileSize.height = remainderHeight;
//			}
//			
//			CGImageRef tileImage = CGImageCreateWithImageInRect(fullImage,
//																(CGRect){{x*size.width, y*size.height},
//																	tileSize});
//			NSData *imageData = UIImagePNGRepresentation([UIImage imageWithCGImage:tileImage]);
//			
//			CGImageRelease(tileImage);
//			
//			NSString *path = [NSString stringWithFormat:@"%@/%@%d_%d.png",
//							  directoryPath, prefix, x, y];
//			[imageData writeToFile:path atomically:NO];
//		}
//	}
//}
//
//
//#import <AppKit/AppKit.h>
//int main(int argc, const char * argv[]) {
//	@autoreleasepool
//	{
//			//handle incorrect arguments
//		if (argc < 2) {
//			NSLog(@"TileCutter arguments: inputfile");
//			return 0; }
//			//input file
//		NSString *inputFile = [NSString stringWithCString:argv[1] encoding:NSUTF8StringEncoding];
//			//tile size
//		CGFloat tileSize = 256; //output path
//		NSString *outputPath = [inputFile stringByDeletingPathExtension];
//			//load image
//		NSImage *image = [[NSImage alloc] initWithContentsOfFile:inputFile];
//		NSSize size = [image size];
//		NSArray *representations = [image representations]; if ([representations count])
//		{
//			NSBitmapImageRep *representation = representations[0]; size.width = [representation pixelsWide];
//			size.height = [representation pixelsHigh];
//		}
//		NSRect rect = NSMakeRect(0.0, 0.0, size.width, size.height); CGImageRef imageRef = [image CGImageForProposedRect:&rect
//																												 context:NULL hints:nil];
//			//calculate rows and columns
//		NSInteger rows = ceil(size.height / tileSize); NSInteger cols = ceil(size.width / tileSize);
//			//generate tiles
//		for (int y = 0; y < rows; ++y) {
//			for (int x = 0; x < cols; ++x) {
//					//extract tile image
//				CGRect tileRect = CGRectMake(x*tileSize, y*tileSize, tileSize, tileSize);
//				CGImageRef tileImage = CGImageCreateWithImageInRect(imageRef, tileRect);
//					//convert to jpeg data
//				NSBitmapImageRep *imageRep = [[NSBitmapImageRep alloc] initWithCGImage:tileImage];
//				NSData *data = [imageRep representationUsingType: NSJPEGFileType properties:nil];
//				CGImageRelease(tileImage);
//					//save file
//				NSString *path = [outputPath stringByAppendingFormat: @"_%02i_%02i.jpg", x, y];
//				[data writeToFile:path atomically:NO]; }
//		} }
//	return 0; }
//

File: UIKit and Animations/WWDC 2010/Designing Apps with Scroll Views/SampleCode/LargeImageDownsizing/LargeImageSwift/LargeImageSwift/Tiling.swift
//
//  Tiling.swift
//  LargeImageSwift
//
//  Created by Seyed Samad Gholamzadeh on 1/12/18.
//  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.
//

import Foundation
import UIKit


func makeTiles(for image: UIImage, to directoryURL: URL, usingPrefix prefix: String, completion: (Bool) -> ()) {
	var scale: CGFloat = 0.125
	while scale <= 1 {
		saveTiles(for: image, inScale: scale, to: directoryURL, usingPrefix: prefix) { (finished) in
			if !finished {
				completion(finished)
				return
			}
			scale *= 2
		}
	}
}
func saveTiles( of size: CGSize = CGSize(width: 256.0, height: 256.0), for image: UIImage, inScale scale: CGFloat, to directoryURL: URL, usingPrefix prefix: String, completion: (Bool) -> ()) {
	
	var image = image
	
	let imageWidth = CGFloat(image.cgImage!.width)
	let imageHeight = CGFloat(image.cgImage!.height)

	// Create a low res image representation of the image to display before the TiledImageView
	// renders its content.
	let imageRect = CGRect(origin: .zero, size: CGSize(width: imageWidth*scale, height: imageHeight*scale))

	if scale != 1 {
		UIGraphicsBeginImageContext(imageRect.size)
		let context = UIGraphicsGetCurrentContext()
		
		context?.saveGState()
		context?.translateBy(x: 0, y: imageRect.height)
		context?.scaleBy(x: 1, y: -1)

		context?.draw(image.cgImage!, in: imageRect)
		context?.restoreGState()
		image = UIGraphicsGetImageFromCurrentImageContext()!
		UIGraphicsEndImageContext()
	}
	
	let cols = imageRect.width/size.width
	let rows = imageRect.height/size.height
	
	var fullColomns = floor(cols)
	var fullRows = floor(rows)
	
	let remainderWidth = imageRect.width - fullColomns*size.width
	let remainderHeight = imageRect.height - fullRows*size.height
	
	if cols > fullColomns { fullColomns += 1 }
	if rows > fullRows { fullRows += 1 }
	
	let fullImage = image.cgImage!

	for row in 0..<Int(fullRows) {
		for col in 0..<Int(fullColomns ){
			var tileSize = size
			if col + 1 == Int(fullColomns) && remainderWidth > 0 {
				// Last Column
				tileSize.width = remainderWidth
			}
			if row + 1 == Int(fullRows) && remainderHeight > 0 {
				// Last Row
				tileSize.height = remainderHeight
			}
			
			let tileImage = fullImage.cropping(to: CGRect(origin: CGPoint(x: CGFloat(col)*size.width, y: CGFloat(row)*size.height), size: tileSize))!
			let imageData = UIImagePNGRepresentation(UIImage(cgImage: tileImage))
			
			let tileName = "\(prefix)_\(Int(scale*1000))_\(col)_\(row).png"
			let url = directoryURL.appendingPathComponent(tileName)
			do {
				try imageData!.write(to: url)
			}
			catch {
				print(error)
				let finished = false
				completion(finished)
				break
			}
		}
	}
	
	let finished = true
	completion(finished)

}

